packages/block-editor/src/components/use-block-sync/block-manager.js

```diff
@@ -0,0 +1,81 @@
+/**
+ * External dependencies
+ */
+import { v4 as uuid } from 'uuid';
+
+/**
+ * WordPress dependencies
+ */
+import { useEffect, useRef } from '@wordpress/element';
+
+function createBlockClientIdManager( shouldMapBlocks ) {
+	const DIRECTION_IN = 'in';
+	const DIRECTION_OUT = 'out';
+	const flipDirection = ( direction ) => {
+		return direction === DIRECTION_IN ? DIRECTION_OUT : DIRECTION_IN;
+	};
+
+	const clientIdMap = { [ DIRECTION_IN ]: {}, [ DIRECTION_OUT ]: {} };
+
+	const mapBlocks = ( blocks, direction ) => {
+		if ( ! Array.isArray( blocks ) || ! shouldMapBlocks ) {
+			return blocks;
+		}
+
+		return blocks.map( ( block ) => ( {
+			...block,
+			innerBlocks: mapBlocks( block.innerBlocks, direction ),
+			clientId: mapBlockId( block.clientId, direction ),
+		} ) );
+	};
+
+	const mapBlockId = ( blockClientId, direction ) => {
+		// Avoid running map code on non-string clientIds. There appear to be times
+		// where this function can be called with `blockClientId = {}`
+		if ( typeof blockClientId !== 'string' || ! shouldMapBlocks ) {
+			return blockClientId;
+		}
+
+		if ( ! clientIdMap[ direction ][ blockClientId ] ) {
+			const newId = uuid();
+			clientIdMap[ direction ][ blockClientId ] = newId;
+			clientIdMap[ flipDirection( direction ) ][ newId ] = blockClientId;
+		}
+		return clientIdMap[ direction ][ blockClientId ];
+	};
+
+	return {
+		updateOutgoingBlocks: ( blocks ) => mapBlocks( blocks, DIRECTION_OUT ),
+		updateIncomingBlocks: ( blocks ) => mapBlocks( blocks, DIRECTION_IN ),
+		getOutgoingBlockId: ( blockId ) => mapBlockId( blockId, DIRECTION_OUT ),
+		getIncomingBlockId: ( blockId ) => mapBlockId( blockId, DIRECTION_IN ),
+	};
+}
+
+const _instanceReference = {};
+export default function useBlockManager( entityId ) {
+	const manager = useRef(
+		createBlockClientIdManager( !! _instanceReference[ entityId ] )
```

Comment:
I'm noticing that there's quite a bit of logic to create multiple instances of the block client id manager. The more I think about it, the more I wonder if it's necessary. Curious about your opinion Noah. Still mulling it over 🤔 

Everything else looks good 👍

Reply:
I don't know if there is a way around it, because we either need one object shared globally, or one object per component. I think the object per component makes a lot more sense, like here. The very first approach relied on generated instance IDs to separate the global object, but we don't have to do that any more, we can just rely on the single instance of the manager. As you can see, we've been able to get rid of `instanceId` and the "instance manager" concept altogether now. :)

I think the logic for creating manager instances makes at least some sense. Technically, it basically just makes a new instance of the block manager if and only if we start pointing at a different entity source. (because the blocks would be different then.) If we had a global instance, it would basically have to delete stuff and change the hierarchy each time you changed entityIds, but now we can just let it be garbage collected when the manager instance goes away

Reply:
> The very first approach relied on generated instance IDs to separate the global object, but we don't have to do that any more, we can just rely on the single instance of the manager. As you can see, we've been able to get rid of instanceId and the "instance manager" concept altogether now. :)

> I think the logic for creating manager instances makes at least some sense. Technically, it basically just makes a new instance of the block manager if and only if we start pointing at a different entity source.

These are really, really good points! I agree with everything you've said.

I gave it more thought, and I realized that there might be one more limitation to this approach. Would there ever be a scenario where we would need to access and compare block ids and client ids across different BlockClientIdManager instances? I can't think of a situation, which makes me inclined to move forward with what we have now.

Reply:
I agree in that I can't think of a use case for that. I think if any feature did need that, it might need to refactor this anyways